benches/dapol.rs

```diff
@@ -1,175 +1,250 @@
-        group.bench_function(BenchmarkId::new("padding", tree_height), |bench| {
-            bench.iter_batched(
-                || {
-                    // generate a proof
-                    let item_idx = rng.sample(item_range);
-                    let tree_index = &items[item_idx].0;
-                    (item_idx, dapol.generate_proof(tree_index).unwrap())
-                },
-                |(item_idx, proof)| {
-                    // time proof verification
-                    proof.verify(&dapol.root(), &items[item_idx].1.get_proof_node())
-                },
-                BatchSize::SmallInput,
-            );
-        });
+    thread_counts.push(max_thread_count);
+
+    let mut ndm_smt = Option::<NdmSmt>::None;
+
+    e.advance().unwrap();
+
+    for h in TREE_HEIGHTS.into_iter() {
+        e.advance().unwrap();
+
+        for t in thread_counts.iter() {
+            e.advance().unwrap();
+
+            for u in NUM_USERS.into_iter() {
+                e.advance().unwrap();
+
+                let max_users_for_height = 2_u64.pow((h - 1) as u32);
+
+                if u > max_users_for_height {
+                    break;
+                }
+
+                let tup: (Height, MaxThreadCount, u64) =
+                    (Height::from(h), MaxThreadCount::from(*t), u);
+
+                // tree build compute time
+                group.bench_function(
+                    BenchmarkId::new(
+                        "build_tree",
+                        format!("{:?}/{:?}/NUM_USERS: {:?}", &tup.0, &tup.1, &tup.2),
+                    ),
+                    |bench| {
+                        bench.iter(|| {
+                            e.advance().unwrap();
+                            ndm_smt = Some(setup::build_ndm_smt(tup.clone()));
+                        });
+                    },
+                );
+
+                // tree build memory usage
+                let alloc = alloc.read().unwrap();
+                let act = act.read().unwrap();
+                let res = res.read().unwrap();
+
+                let mem_usage = MemoryUsage {
+                    allocated: setup::bytes_as_string(alloc),
+                    active: setup::bytes_as_string(act),
+                    resident: setup::bytes_as_string(res),
+                };
+
+                // tree build file size
+                let tree_build_file_size = setup::serialize_tree(
+                    ndm_smt.as_ref().expect("Tree not found"),
+                    PathBuf::from("./target"),
+                );
+
+                let tree_build = Metrics {
+                    variable: Variable::TreeBuild,
+                    mem_usage,
+                    file_size: tree_build_file_size,
+                };
+
+                println!("\n{:?}\n", tree_build);
+
+                let alloc = stats::allocated::mib().unwrap();
+                let act = stats::active::mib().unwrap();
+                let res = stats::resident::mib().unwrap();
```

Comment:
You don't need to reset these like this again. Just advance the epoch and read from the already-created variable.

Reply:
This relates to above